File: web/src/types/grade.ts
/**
 * Grade-related types.
 *
 * Wire types (matching the FastAPI response shape) keep snake_case to stay
 * consistent with the rest of `web/src/types/`. Component-prop types use
 * camelCase per JS convention; conversion happens at the call site.
 */

// ---------- Wire types (mirror app/schemas/grade.py) -----------------------

export interface GradeBrief {
  composite_score: number;
  letter_grade: string;
  worth_the_fee: boolean;
  is_complete: boolean;
}

export interface GradeComponent {
  score: number;
  explanation: string;
}

export interface GradeDetail {
  transfer_id: number;
  composite_score: number;
  letter_grade: string;
  worth_the_fee: boolean;
  is_complete: boolean;
  components: Partial<Record<ComponentKey, GradeComponent>>;
  model_version: string;
  graded_at: string;
}

export type ComponentKey = "minutes" | "production" | "value_trajectory" | "financial_return";

export interface RankedTransfer {
  transfer_id: number;
  player_name: string;
  player_id: number;
  from_club_name: string;
  to_club_name: string;
  to_club_country: string;
  fee_eur: number;
  composite_score: number;
  letter_grade: string;
  worth_the_fee: boolean;
  is_complete: boolean;
  transfer_window: string;
  position_group: string | null;
  summary: string;
}

export interface TopGradesResponse {
  category: string;
  total: number;
  transfers: RankedTransfer[];
}

export interface ClubBestWorst {
  transfer_id: number;
  player_id: number;
  player_name: string;
  composite_score: number;
  letter_grade: string;
  fee_eur: number;
}

export interface ClubTimeline {
  transfer_window: string;
  average_grade: number;
  transfer_count: number;
}

export interface ClubReportCard {
  club_id: number;
  club_name: string;
  overall_average_grade: number;
  overall_letter_grade: string;
  total_graded_transfers: number;
  total_spend_eur: number;
  best_transfer: ClubBestWorst | null;
  worst_transfer: ClubBestWorst | null;
  grade_distribution: Record<string, number>;
  timeline: ClubTimeline[];
}

export interface ClubComparisonResponse {
  clubs: ClubReportCard[];
}

export interface PlayerGradeSummary {
  average_grade: number;
  total_graded_transfers: number;
  best_grade: ClubBestWorst | null;
  worst_grade: ClubBestWorst | null;
}

export interface ScoringInfoResponse {
  version: string;
  scored_at: string;
  total_scored: number;
  component_weights: Record<string, number>;
  calibrated_rates: Record<string, Record<string, number>>;
  grade_distribution: Record<string, number>;
}

// ---------- Query params for /grades/top -----------------------------------

export interface TopGradesParams {
  category?: "best" | "worst";
  limit?: number;
  offset?: number;
  positionGroup?: string | null;
  countryId?: number | null;
  feeMin?: number | null;
  feeMax?: number | null;
  windowStart?: string | null;
  windowEnd?: string | null;
  completedOnly?: boolean;
}

File: api/app/schemas/grade.py
"""Grade-related response schemas.

`GradeBrief` is the abbreviated form embedded in transfer rows across the existing
endpoints. `GradeDetail` is the full breakdown returned by `/transfers/{id}/grade`
on hover. The other classes back the new `/grades/*` endpoints.
"""

from datetime import datetime

from pydantic import BaseModel


class GradeBrief(BaseModel):
    """Abbreviated grade attached to transfer rows in country/player/network endpoints."""

    composite_score: float
    letter_grade: str
    worth_the_fee: bool
    is_complete: bool


class GradeComponent(BaseModel):
    """One of the four scoring components in the breakdown tooltip."""

    score: float
    explanation: str


class GradeDetail(BaseModel):
    """Full grade with per-component scores + explanations + version metadata."""

    transfer_id: int
    composite_score: float
    letter_grade: str
    worth_the_fee: bool
    is_complete: bool
    components: dict[str, GradeComponent]
    model_version: str
    graded_at: datetime


class RankedTransfer(BaseModel):
    """One row in the best/worst rankings list."""

    transfer_id: int
    player_name: str
    player_id: int
    from_club_name: str
    to_club_name: str
    to_club_country: str
    fee_eur: int
    composite_score: float
    letter_grade: str
    worth_the_fee: bool
    is_complete: bool
    transfer_window: str
    position_group: str | None
    summary: str


class TopGradesResponse(BaseModel):
    """Paginated rankings response."""

    category: str
    total: int
    transfers: list[RankedTransfer]


class ClubBestWorst(BaseModel):
    """Single best- or worst-graded transfer for a club's report card."""

    transfer_id: int
    player_id: int
    player_name: str
    composite_score: float
    letter_grade: str
    fee_eur: int


class ClubTimeline(BaseModel):
    """One point in a club's grade-over-time line chart."""

    transfer_window: str
    average_grade: float
    transfer_count: int


class ClubReportCard(BaseModel):
    """Per-club aggregate grade statistics for the comparison view."""

    club_id: int
    club_name: str
    overall_average_grade: float
    overall_letter_grade: str
    total_graded_transfers: int
    total_spend_eur: int
    best_transfer: ClubBestWorst | None
    worst_transfer: ClubBestWorst | None
    grade_distribution: dict[str, int]
    timeline: list[ClubTimeline]


class ClubComparisonResponse(BaseModel):
    """Up to 3 clubs' report cards side-by-side."""

    clubs: list[ClubReportCard]


class ScoringInfoResponse(BaseModel):
    """Active scoring configuration + grade-distribution snapshot for diagnostics."""

    version: str
    scored_at: datetime
    total_scored: int
    component_weights: dict[str, float]
    calibrated_rates: dict[str, dict[str, float]]
    grade_distribution: dict[str, int]


class PlayerGradeSummary(BaseModel):
    """Aggregated grade summary attached to a Player detail response."""

    average_grade: float
    total_graded_transfers: int
    best_grade: ClubBestWorst | None
    worst_grade: ClubBestWorst | None
